was14rdma

was4 opened the pull request:
See extensive commit header for details. Authentication not tested, but mostly simple symbol changes by inspection. After discussion with Marcus, leaving more substantive GSS changes until another pass.

mattbenjamin commented on ntirpc/rpc/auth_inline.h:
```diff
@@ -156,10 +153,6 @@ inline_auth_decode(XDR *xdrs, struct opaque_auth *oa, int32_t *buf)
 static inline bool
 inline_auth_free(XDR *xdrs, struct opaque_auth *oa)
 {
-	if (oa->oa_base != NULL) {
```
@was4  cool

mattbenjamin commented on ntirpc/rpc/rpc_msg.h:
```diff
@@ -103,6 +102,9 @@ struct accepted_reply {
 #define ar_vers  ru.AR_versions
+
+	/* after union to avoid (rare) corruption by rejected_reply */
+	struct opaque_auth ar_verf;
```
ok

mattbenjamin commented on ntirpc/rpc/rpc_msg.h:
```diff
@@ -139,11 +141,6 @@ struct reply_body {
  */
 struct call_body {
 	rpcvers_t cb_rpcvers;	/* must be equal to two */
-	rpcprog_t cb_prog;
```
ok

mattbenjamin commented on src/svc.c:
```diff
@@ -230,39 +230,13 @@ svc_init(svc_init_params *params)
-{
-	struct rpc_msg *msg = mem_alloc(sizeof(struct rpc_msg));
-
-	TAILQ_INIT_ENTRY(msg, msg_q);
```
@was4 so where has this moved?  this is needed for bi-directional

was4 replied:
It hasn't moved. It is never used anywhere. I'm trying to figure out what you meant, but see nothing in the rpc_call_ctx other than the rb tree.

In the ganesha nfs_core.h, there is a struct glist_head req_q; is that what you mean?

mattbenjamin replied:
I'm worried about the rpc_msg itself.  In client call, iirc, we allocate one, but when bi-directional, we get a pointer to one allocated in svc_recv (or its caller?) and work with it.

was4 replied:
Ah. Now copied. No alloc/free.

> 		ctx->cc_msg = *msg;	/* and stash reply header */

mattbenjamin replied:
yeah, ok, that's what I was looking for

mattbenjamin reviewed (commented):
@was4 I strongly approve of removing the allocation, but at the same time I introduced it to cope with bi-directional calls where a message that svc_recv() decoded might need to be returned (and in the model as it was) freed by clnt_call().  I'm guessing that changes to clnt_call() are needed?

mattbenjamin commented:
@was4 see above--I think you're on the right track, but I think that as is, this will break clnt_call w/bi-directional; presumably can be adapted--I agree, I'd rather not impose another allocation of rpc_msg going forward

mattbenjamin commented:
updated

Matt

----- Original Message -----
> From: "was4" <notifications@github.com>
> To: "nfs-ganesha/ntirpc" <ntirpc@noreply.github.com>
> Cc: "Matt Benjamin" <mbenjamin@redhat.com>, "Mention" <mention@noreply.github.com>
> Sent: Friday, December 16, 2016 2:46:19 PM
> Subject: Re: [nfs-ganesha/ntirpc] was14rdma (#30)
> 
> @was4 requested your review on: nfs-ganesha/ntirpc#30 was14rdma.
> 
> --
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub:
> https://github.com/nfs-ganesha/ntirpc/pull/30#event-897162402

-- 
Matt Benjamin
Red Hat, Inc.
315 West Huron Street, Suite 140A
Ann Arbor, Michigan 48103

http://www.redhat.com/en/technologies/storage

tel.  734-821-5101
fax.  734-769-8938
cel.  734-216-5309

dang reviewed (approved):
Seems fine to me, but I'm not an expert in this code.

was4 commented:
Pushed a new version with the conflict (of an hour ago) resolved.

Commit:
Embed rpc_msg in svc_req and rpc_call_ctx

Avoids separate alloc/free -- this is much more efficient.

Also removes remaining alloc/free of authentication cred and verf;
bugfix: these were freed based on non-NULL oa_base, even though many
instances pointed to arrays in rpc_msg.

Don't use stack rpc_msg. These could disappear during RDMA processing.

Remove obsolete pointer rq_clntcred; bugfix: it was never updated, was
pointing at freed memory after rpc_msg was freed.

Remove duplicate variables formerly copied from rpc_msg (and back):
  rq_prog, rq_vers, rq_proc, rq_cred, rq_xid, rq_verf.

Instead, move any call_body equivalents out of the union, so that they
aren't overwritten and can be used for authentication and logging.
This process was mostly mechanical, so no code changes should occur.

Also move ar_verf after its union to avoid possible corruption.
(Probably an overabundance of caution.)

Replace oa_base pointer with its oa_body, reducing alloc/free thrashing
in auth code at the initialization expense of copying the zeroed body.

Replace inline_xdr_bytes() with inline_xdr_opaques() that does not
alloc/free. It was rarely called.

Remove inline_xdr_netobj(). It was never called.

Rename a couple of ambiguous deliberately misspelled symbols, as they
were still too long for some lines.  Now RPCM_ack and RPCM_rej.

Remove redundant SCVXPRT parameter(s).

Remove redundant authentication parameter(s).

Signed-off-by: William Allen Simpson <william.allen.simpson@redhat.com>
